File: gc-commons/gc-common/src/main/java/com/gc/common/utils/TreeUtils.java
package com.gc.common.utils;

import com.gc.common.model.Tree;
import com.google.common.collect.Lists;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.List;

/**
 * 树形工具类
 * @author shizhongming
 * 2020/1/8 9:02 下午
 */
public class TreeUtils {

    @Nullable
    public static <T> Tree<T> build(@Nullable List<Tree<T>> nodes) {
        if (nodes == null) {
            return null;
        }
        final List<Tree<T>> topNodes = Lists.newArrayList();
        for (var children : nodes) {
            final var pid = children.getParentId();
            if (pid == null || "0".equals(pid)) {
                topNodes.add(children);
                continue;
            }
            for (var parent : nodes) {
                final var id = parent.getId();
                if (id != null && id == pid) {
                    parent.getChildren().add(children);
                    children.setHasParent(true);
                    parent.setHasChildren(true);
                }
            }
        }
        var root = new Tree<T>();
        if (topNodes.size() == 1) {
            root = topNodes.get(0);
        } else {
            root.setId(-1);
            root.setParentId("");
            root.setHasParent(false);
            root.setHasChildren(true);
            root.setChildren(topNodes);
            root.setText("顶级节点");
        }
        return root;
    }

    /**
     * 构建树形列表
     * @param nodes 树形数据
     * @param idParam 上级ID
     * @param <T>
     * @return
     */
    @NotNull
    public static <T> List<Tree<T>>  buildList(@Nullable List<Tree<T>> nodes, @NotNull Serializable idParam) {
        if (nodes == null) {
            return List.of();
        }

        final List<Tree<T>> topNodes = Lists.newArrayList();
        for (var children : nodes) {
            final var pid = children.getParentId();
            if (pid == null || idParam == pid) {
                topNodes.add(children);
                continue;
            }
            for (var parent : nodes) {
                val id = parent.getId();
                if (id != null && id == pid) {
                    parent.getChildren().add(children);
                    children.setHasParent(true);
                    parent.setHasChildren(true);
                }
            }
        }
        return topNodes;
    }
}

File: smart-commons/smart-common/src/main/kotlin/com/smart/common/utils/TreeUtils.kt
package com.smart.common.utils

import com.smart.common.model.Tree
import java.util.*

/**
 * 属性数据工具类
 * @author ming
 * 2019/6/12 上午9:55
 */
object TreeUtils {
    /**
     * 构建树形
     */
    @JvmStatic
    fun <T> build(nodes: List<Tree<T>>?): Tree<T>? {
        if (nodes == null) {
            return null
        }
        val topNodes = ArrayList<Tree<T>>()
        for (children in nodes) {
            val pid = children.parentId
            if (pid == null || "0" == pid) {
                topNodes.add(children)
                continue
            }
            for (parent in nodes) {
                val id = parent.id
                if (id != null && id == pid) {
                    parent.children.add(children)
                    children.isHasParent = true
                    parent.setChildren(true)
                    continue
                }
            }
        }
        var root = Tree<T>()
        if (topNodes.size == 1) {
            root = topNodes[0]
        } else {
            root.id = "-1"
            root.parentId = ""
            root.isHasParent = false
            root.setChildren(true)
            root.isChecked = true
            root.children = topNodes
            root.text = "顶级节点"
            val state = HashMap<String, Any>(16)
            state["opened"] = true
            root.state = state
        }
        return root
    }

    /**
     * 构建树形列表
     */
    @JvmStatic
    fun <T> buildList(nodes: List<Tree<T>>?, idParam: String): List<Tree<T>> {
        if (nodes == null) {
            return emptyList()
        }
        val topNodes = ArrayList<Tree<T>>()
        for (children in nodes) {
            val pid = children.parentId
            if (pid == null || idParam == pid) {
                topNodes.add(children)
                continue
            }
            for (parent in nodes) {
                val id = parent.id
                if (id != null && id == pid) {
                    parent.children.add(children)
                    children.isHasParent = true
                    parent.setChildren(true)
                    continue
                }
            }
        }
        return topNodes
    }

    /**
     * 反转父子关系
     */
    @JvmStatic
    @Deprecated("")
    fun reverseParentChild(parentChildren: Map<String, List<String>>, withSelf: Boolean = true): Map<String, List<String>> {
        var selefParentChildren = parentChildren
        if (withSelf) {
            selefParentChildren = parentChildren.map {
                it.key to it.value.let { value ->
                    val list = mutableListOf<String>()
                    list.add(it.key)
                    list.addAll(value)
                    list
                }
            }.toMap()
        }
        val result = mutableMapOf<String, MutableList<String>>()
        selefParentChildren.forEach { key, list ->
            list.forEach {
                if (result[it] == null) result[it] = mutableListOf()
                result[it]!!.add(key)
            }
        }
        return result
    }
}